concordium-std/src/impls.rs

```diff
@@ -1037,3 +2479,134 @@ impl DeserialCtx for String {
+        source.read_array().map(|set_prefix| StateSet::open(state.clone(), set_prefix))
+    }
+}
+
+impl<T, S> DeserialWithState<S> for StateBox<T, S>
+where
+    S: HasStateApi,
+    T: Serial + DeserialWithState<S>,
+{
+    fn deserial_with_state<R: Read>(state: &S, source: &mut R) -> ParseResult<Self> {
+        let prefix = source.read_array()?;
+        Ok(StateBox {
+            state_api: state.clone(),
+            inner:     UnsafeCell::new(StateBoxInner::Reference {
+                prefix,
+            }),
+        })
+    }
+}
+
+impl<T: Serialize> Deletable for T {
+    #[inline(always)]
+    fn delete(self) {} // Types that are Serialize have nothing to delete!
+}
+
+impl<T, S> Deletable for StateBox<T, S>
+where
+    T: Serial + DeserialWithState<S> + Deletable,
+    S: HasStateApi,
+{
+    fn delete(mut self) {
+        // replace the value with a dummy one for which drop is a no-op.
+        let inner = mem::replace(
+            &mut self.inner,
+            UnsafeCell::new(StateBoxInner::Reference {
+                prefix: [0u8; 8],
+            }),
+        );
+        let (entry, value) = match inner.into_inner() {
+            StateBoxInner::Loaded {
+                entry,
+                value,
+                ..
+            } => (entry, value),
+            StateBoxInner::Reference {
+                prefix,
+            } => {
+                // we load the value first because it might be necessary to delete
+                // the nested value.
+                // TODO: This is pretty bad for performance, but we cannot specialize the
+                // implementation for flat values.
+                let mut entry = self.state_api.lookup_entry(&prefix).unwrap_abort();
+                let value = T::deserial_with_state(&self.state_api, &mut entry).unwrap_abort();
+                (entry, value)
+            }
+        };
+        self.state_api.delete_entry(entry).unwrap_abort();
+        value.delete()
+    }
+}
+
+impl<T, S> Deletable for StateSet<T, S>
+where
+    S: HasStateApi,
+{
+    fn delete(mut self) {
+        // Statesets cannot contain state types (e.g. StateBox), so there is nothing to
+        // delete, apart from the set itself.
+
+        // Unwrapping is safe when only using the high-level API.
+        self.state_api.delete_prefix(&self.prefix).unwrap_abort();
+    }
+}
+
+impl<K, V, S> Deletable for StateMap<K, V, S>
+where
+    S: HasStateApi,
+    K: Deserial,
+    V: DeserialWithState<S> + Deletable,
+{
+    fn delete(mut self) {
+        // Delete all values pointed at by the statemap. This is necessary if `V` is a
+        // StateBox/StateMap.
+        for (_, value) in self.iter() {
+            value.value.delete()
+        }
+
+        // Then delete the map itself.
+        // Unwrapping is safe when only using the high-level API.
+        self.state_api.delete_prefix(&self.prefix).unwrap_abort();
```

Comment:
```suggestion
    K: Serialize,
    V: Serial + DeserialWithState<S> + Deletable,
{
    fn delete(mut self) {
        self.clear();
```

Reply:
nitpick: I am not sure about the implications of the changes in the trait bounds but it's worth checking. The compiler did not complain.